Add platform and hosting check, add hosting error message, continue anyway path for WPCOM sites

Imran92 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/94915

## Proposed Changes

* Added functionality to fetch platform and hosting info for source site.
* Added error message for displaying if site's already on WordPress.com.
* Also, we won't need any further checks to implement to direct users to either the survey page or other platform importing pages. We'll just need to direct them when we have the pages ready.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* To let the users know that they are already on WordPress.com

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Proceed through the migration flow. Make sure to give input to a site that's already on WordPress.com
- On the "How do you want to migrate?" step, choose "Do it for me".
- In the credentials form, fill the fields and click on Continue button. Check for the following cases-

#### Site already on WordPress.com
- Make sure you see an error below the source URL field to remind you that the site is already on WordPress.com
- Make sure the Continue button has changed to "Continue anyways". Click that button
- Make sure your request completes as expected and you are taken to the next page

#### WP site, but not on WordPress.com
- Now repeat the whole flow with a site that's not on wordpress.com, but is still a valid Wordpress site. Once you click on the Continue button, you shouldn't see any error (if your creds are valid) and your request should complete as expected.

#### Non-WP site
- Now repeat the whole process with a site that's not a WordPress site at all. Once you click on the Continue button, you shouldn't see similar validation and "Continue anyways" behavior of the button as before

https://github.com/user-attachments/assets/5d32eabb-8d60-4f52-8c9e-1c738508a9a5

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Imran92 committed:
Add mapping functionality for platform check error

Imran92 committed:
Implement URL analyzer check to find hosting

Imran92 committed:
Implement URL analyzer check to find hosting

Imran92 committed:
Show loader info when analyzing site URL

Imran92 committed:
Add test to make sure analyzer check and WPCOM error message state works

Imran92 committed:
Add test to make sure analyzer check and WPCOM error message state works

Imran92 committed:
Add test to make sure analyzer check and WPCOM error message state works

gabrielcaires reviewed (changes requested):
@Imran92  Some thoughts.

 We should not wait for the user to click submit to show the validation; it is important because the process to get the site credentials probably requires some time, and when the site is wpcom, the credentials don't matter more, causing a lot of frustration.

I think it should solve in two parts: 

**1.** When the user arrives on the step with form=WPCOMSITE.COM.  (Most of the cases)
The form will start the validation automatically without wait the user click on the button. 

**2. The step is loaded without the URL on the FROM PARAM** 
As soon as the user types the URL and selects another field or tries to click on the button (blur event), we disable the button until have a validation.

OR

If we want to keep the current approach of waiting the user click on the button, we have use the following changes:

1. Don't use `useAnalyzeUrlQuery`  hook.
2. Update the  `submitHandler` to be async 
3. Make the call directly on the hook. 
4. Use the useForm > formState > isSubmitting to capture when the form is still processing the async call, so we can just disable the button using this flag, instead of the `isPending` from the 

Something like:
```
	const submitHandler = async ( data: CredentialsFormData ) => {
		console.log("Validating ....")
               try {
		 await analyzeURL(data.from_site)  
                  await requestAutomatedMigration() // Requires to use the mutateSync version, instead of the mutate as I demostred above
                 reset();
		clearErrors();
               } catch (error) {
                   /// code to set custom errors
               }
	};

```

We can use the async version of the mutation.
```
const {
	...
		mutateAsync: requestAutomatedMigration,
	...
	} = useSiteMigrationCredentialsMutation();
```



I think the main issue here is we are trying to be sometimes reactive and sometimes in sync with the `requestAutomatedMigration` call.

fditrapani commented:
I'll wait to test this after you have a chance to digest Gabriel's feedback but one small note is to make the url field editable after we throw an error or maybe never disable it at all?

Imran92 commented:
Thanks for sharing your thoughts @gabrielcaires ! These are very nice inputs.

I'll change the code to validate when typing instead of upon submission. It'll be a simple change I think. I'll also take @fditrapani 's suggestion and make the URL field always editable.

I have just one additional thought about the first point -

> 1. When the user arrives on the step with form=WPCOMSITE.COM. (Most of the cases)
> The form will start the validation automatically without wait the user click on the button.

Not for now, but maybe later we can think about not bringing a user to the credential at all if they put in a URL to a site that's already on WPCOM.

I noticed that we already have a sort of similar behavior in the /move page, but not in the site identify step of the flow. Maybe if we detect in the site identify step that the site is already on WPCOM, we can take the user to the survey page directly from there
<img width="849" alt="image" src="https://github.com/user-attachments/assets/5d13725e-1436-458d-adff-0e70b42cca6e">

fditrapani commented:
> Not for now, but maybe later we can think about not bringing a user to the credential at all if they put in a URL to a site that's already on WPCOM.

I think that's a good idea!

gabrielcaires commented:
> I'll change the code to validate when typing or upon loading instead of submission. 

Nice! 
Do you have any thoughts about my concerns with the code? This change is making the form excessively complex.

Imran92 committed:
Enable URL field, check site info on typing

Imran92 committed:
Remove unnecessary param from site address

Imran92 committed:
Remove unnecessary test

Imran92 committed:
Remove unnecessary dep

Imran92 commented:
Hi @gabrielcaires and @fditrapani 👋 

I have implemented the following changes as per discussion
- Perform the site platform check on load if the source URL is already available, else perform the check on change event of the Site Address field
- Make the Site Address field always editable, pre-populate it with URL if it's available. Allowing users the ability to change it
- Made changes in the code to streamline the logic and make it simpler

Also, I will create an issue to disable the requirement of Username and Password field to be non-empty if we detect that the site is on wpcom already. But it'll require backend changes as well. So it'll be better to handle it separately

fditrapani commented:
@Imran92 I'm having trouble getting this running locally. Would you mind posting another video with the updates?

Imran92 commented:
Hi @fditrapani 👋 I've updated the description with the latest video. Thank you!

fditrapani commented:
Thanks for posting the updated video. Overall this is looking great. 

One small request, when we're doing inline validation of the URL field, the other fields get blocked. The button says verifying credentials. It feels strange to block up the entire UI while we're checking the UI. Can we perhaps do that in the background without blocking up the username + password field? The disabled fields make more sense when you submit the entire form.

Imran92 committed:
Disable only site address button during site info verification

Imran92 committed:
Disable only site address button during site info verification

Imran92 committed:
Disable only site address button during site info verification

Imran92 committed:
Disable only site address button during site info verification

fditrapani commented:
Thanks for making the changes @Imran92. It's getting close! The inline validation shouldn't block any other parts of the UI while it's checking. Without initiating the user submission, it looks like something unintentional is happening. We can still run the inline field validation but without blocking the continue or URL fields while it's happening? I think it's fine to block the UI after someone presses the continue button but otherwise, we should only show the error message on the field if we validate and something is wrong with the URL.

Imran92 committed:
Check site platform on submit

Imran92 committed:
Check site platform on submit

Imran92 committed:
Check site platform on submit

Imran92 committed:
Check site platform on submit

Imran92 commented:
Hi @fditrapani 👋 Thanks for checking it! I have updated it check only on submission to simplify both the code and the behavior. I have updated the description with the latest video

Imran92 committed:
Remove references of useAnalyzeUrlQuery from test

fditrapani reviewed (approved):
Thanks for the updates @Imran92. I didn't review the code but this looks great now. It lines up much better with how people would expect a form to work :)

gabrielcaires reviewed (approved):
LGTM 🚢

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16898976).
